Add regex query support to wildcard field (approach 2)

markharwood opened the pull request:
This is the second cut at extracting ngrams from automata for use in the approximation query. 

Unlike https://github.com/elastic/elasticsearch/pull/54946 this PR uses FiniteStringsIterator on a simplified form of automaton. The automaton simplification is achieved by two means:
1) No expansions -rather than generating complex automata a single null character is used in place of * type expansions. This make the job of extracting runs of concrete characters simple via FiniteStringsIterator.  This was acheived by forking Lucene's RegExp and WildcardQuery's logic to create automatons. 
2) Unified case - to avoid overly-complex paths and ngram queries for mixed-case searches like `[Ee][Nn][Cc][Oo][Dd]` I have opted to lower-case the search inputs and the ngram index.

Fuzzy queries remain a problem because they generate automatons that are too complex to map to an efficient BooleanQuery on the ngram index so are not supported.

Closes #54725

markharwood committed:
Adds full equivalence for keyword field to the wildcard field. Regex, fuzzy, wildcard and prefix queries are all supported.
All queries use an approximation query backed by an automaton-based verification queries.

Closes #54275

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,187 +229,175 @@ public WildcardFieldType clone() {
+            failIfNotIndexed();
+            RegExp regex = new RegExp(value, flags);
+            Automaton automaton = regex.toAutomaton(maxDeterminizedStates);
+            ApproximateRegExp ngramRegex = new ApproximateRegExp(toLowerCase(value), flags);
```
I don't think we should lowercase eagerly here. Lowercasing can happen in the `ApproximateRegExp` directly but I wonder if this should be tackled in a follow up ?

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,187 +229,175 @@ public WildcardFieldType clone() {
+                // determine a common section all the paths shared in order to simplify.
+                // So, we simplify up-front by:
+                // 1) replacing all repetitions eg (foo)* with a single invalid char in the regex string used to build the automaton
+                // 2) lowercasing all concrete values so searches like [Ee][Nn][Cc][Oo][Dd][Ee][Dd] don't fork myriad paths
```
I'd like to see the solution without lowercasing the ngram index. We can try to optimize lowercased search in a follow up but for now we should ensure that we handle all types of automaton without relying on normalization ?

markharwood replied:
I think I need to understand better how an FSI approach can be made to work with an approach that tries to find articulation points in the automaton. The former relies on FSI to chase down all the possible paths while the latter takes direct control of exploring the graph paths to understand the branching (like Nik's code). 
If we try to make life simple by using FSI only then some up-front case normalisation is useful to avoid searches like the `[Pp][Oo][Ww][Ee][Rr][Ss][Hh][Ee][Ll][Ll]` example from [this blog](https://socprime.com/en/blog/elastic-for-security-analysts-part-1-searching-strings/) blowing the BooleanQuery limits.

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,187 +229,175 @@ public WildcardFieldType clone() {
+                tokenizer.reset();
+                boolean takeThis = true;
+                // minimise number of terms searched - eg for "12345" and 3grams we only need terms
+                // `123` and `345` - no need to search for 234. We take every other ngram.
```
Should we skip 2 positions instead of 1 when there are enough characters ? `123456` should produce `123` and `456` rather than `123`, `345`, `456` ?

markharwood replied:
The original thinking was this might help keep them more position-based by having the middle term act as a bridge

jimczi replied:
I see, thanks.

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,187 +229,175 @@ public WildcardFieldType clone() {
+        
+
+        private Query createApproximationQueryFromAutomaton(Automaton simplifiedAutomaton) {
+            FiniteStringsIterator iterator = new FiniteStringsIterator(simplifiedAutomaton);
```
I think that's ok for a first iteration but we should look at dividing the boolean query around the automaton's [articulation points](https://en.wikipedia.org/wiki/Biconnected_component). That could optimize a lot the boolean query that we produce on complex automaton.

markharwood replied:
Doesn't that take us back to Nik's code and exploring the transition points in the automaton to understand the graph? I'm not sure how FiniteStringsIterator works in conjunction with that sort of logic.

jimczi replied:
We have an example in Lucene in `QueryBuilder#analyzeGraphBoolean`. I don't think we should do that in the first iteration though so let's assume that we have to visit all the paths for now. 
However I am thinking of one thing that we could try to optimize in this pr, when we have a few transitions that have the same target. For instance `[aA][bB][cD]` that would create 8 paths, we could instead allow each path to contain a few unicode points per positions (2 maybe 3 max) and fork `FiniteStringIterator` to return optimized `IntsRef[][]` ?

markharwood replied:
I'm not sure why we don't normalize case given:

- case is likely to be the biggest cause of logic-branching (based on the existing rules we've seen from the security space) 
- normalising reduces our code complexity and increases the number of expressions we can hope to accelerate
- the ngram search just has to be fast, not 100% correct. Normalising will:
  - dramatically reduce the speed of searches (up to 8 x fewer unique terms)
  - not increase false positives massively for most cases (gut feel on this)

jimczi replied:
Sure but I still think we should consider this change as a follow up. This pr should ensure that we can handle all types of automaton correctly no matter if we lowercase or not. 
Then we can discuss how to handle case insensitive search but that's a different discussions imo.

markharwood replied:
>Then we can discuss how to handle case insensitive search but that's a different discussions imo.

The idea behind normalisation in this context isn't about enabling case insensitivity for end users - it's about optimising the search performance and minimising the complexity of automatons. 
There will inevitably be limits on the number of permutations we can consider and  working with a lower-cased vocabulary will help reduce the numbers.

jimczi replied:
That's assuming that lots of query will contains `[aA][bB]` variations but this could be avoided if we provide a true case-insensitive search. Again, I am not saying we shouldn't do this "optimization" but to me this change would deserve a separate discussion where we'd also see think about exposing the case insensitive search on this field.

jimczi reviewed (commented):
I left some comments but the approach looks good to me.

markharwood commented:
>I don't understand why we cannot extract @&~ in the first regexp and aaa AND bbb in the second one ?

In the first example the regex is for matching everything except terms beginning with 'abc'. Currently we ignore any negatives and focus on creating MUST clauses for the positive elements of a regex.
In the second example I'm not sure what's going on. The RegExp fork I made currently produces an empty automaton and the standard RegExp class produces an automaton with cycles which means we can't use FSI on it directly.

markharwood commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,187 +229,174 @@ public WildcardFieldType clone() {
+        
+        @Override
+        public Query regexpQuery(String value, int flags, int maxDeterminizedStates, RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
```
Is it worth reserving it for patterns that don't produce an approximation query?

jimczi commented:
> In the first example the regex is for matching everything except terms beginning with 'abc'.

I see it now, thanks. 

> In the second example I'm not sure what's going on.

I also missed the intersection that is used `&` so it's ok that we rewrite to an empty automaton. However I wonder if we should have a special treatment for intersection that explicitly skip them (replace right and left with the special char) in the approximate regexp ?

markharwood commented:
@jimczi I've had a rethink on the approach to extracting queries from automatons. The original motivation was the hope that a single implementation could work for all query types (wildcard, regex, fuzzy). Life didn't turn out that simple for a number of reasons:

1) Fuzzy queries were too complex to derive from automaton (too many permutations to represent as query terms). We opted to create ngram queries directly from the search string.
2) We had to fork RegExp and WildcardQuery parsing logic to simplify the automatons we dealt with for regex and wildcard searches. The parsing logic for each class was modified to introduce null characters in place of too-complex expressions like negations or repetitions.

So we either side-stepped automatons completely (fuzzy) or had to create alternative ones to the automatons used in verification by using forks of the regex and wildcard parsing logic.

What I have in this PR I think is much simplified.
1) Wildcards and prefix queries create simple ngram runs directly from the input string
2) Fuzzy queries create ngram queries directly from the input string with `min-should-match` settings that reflect the allowed edit distances and MUST clauses that respect the prefix length settings
3) The ApproximateRegExp fork of RegExp uses the regex parser logic to pull out BooleanQuery and TermQuery objects rather than having an interim step of generating automata. This preserves the original logic of the expression without additional translation steps.

#### Regex complexity
The `[Cc][Aa][Ss][Ee]` type regex queries we know are out there waiting to hit us will be responsible for producing complex ngram queries. These can be translated in either of these ways:
a) Character-wise pairings e.g. `+(C* OR  c*) +(A* OR a*) +(S* OR s*) +(E* OR e*) `
b) Sequence expansion e.g. `cas OR Cas OR cAs OR caS OR CAs OR CAS OR CaS...`

Obviously option b) has limits based on input string length because of the possible permutations. However trimming is not an option - with OR lists dropping any one clause will introduce false negatives which is not allowed. It's an all-or-nothing approach to capturing the logic.
I have implemented option a). This is not great because of all the `x*` wildcard queries required. These are expensive and not very selective.

#### Regex normalizatiom
Given the above, I have opted to use lowercasing on the ngram index and the ApproximateRegExp class will optimise these character sequences so that `[Pp][Oo][Ww][Ee][Rr][Ss][Hh][Ee][Ll][Ll]` killer query becomes the much cheaper and selective `powershell`  which is ngrammed to `+_po +owe +ers +she +ell_ +l__`

markharwood committed:
Formatting tidy and tightened expectations in test results for acceleration queries

markharwood committed:
Unused import

markharwood commented:
Updating this feature comparison table as of this PR.

Feature | keyword | wildcard
--|--|--
Sort by speeds| Fast | Not quite as fast _(see *1)_
Aggregation speeds | Fast | Not quite as fast _(see *1)_
Prefix query speeds (foo*)| Fast | Not quite as fast _(see *2)_
Leading wildcard query speeds on low-cardinality fields (*foo)| Fast| Slower _(see *3)_
Leading wildcard query speeds on high-cardinality fields (*foo)| Terrible| Much faster
Term query. full value match (foo) | Fast  | Not quite as fast _(see *2)_
Fuzzy query. | Y  _(if allow expensive queries enabled)_| Y
Regex query. | Y _(if allow expensive queries enabled)_  | Y
Range query. | Y _(if allow expensive queries enabled)_  | N
Disk costs for mostly unique values | high | lower
Disk costs for mostly identical values | low | medium
Max character size for a field value|256 for default JSON string mappings, 32,766 Lucene max|unlimited

* 1: somewhat slower as doc values retrieved from compressed blocks of 32
* 2: somewhat slower because approximate matches with ngrams need verification
* 3: keyword field visits every unique value but wildcard field visits every doc that is similar to value

markharwood committed:
Simplified BooleanQueries and made a* ngram queries prefix rather than wildcard.
Updated tests with simpler syntax and documented regexes that we’d like to improve on, showing current suboptimal queries and the future form we’d like to see.

markharwood commented:
@jpountz @jimczi 
In the interests of time this is something that can be reviewed now as I have no major changes planned.
I'm hoping to add acceleration for `.*` type queries so we skip the verification phase and revert to a plain MatchAllDocs query. This logic is proving a little tricky to add for more deeply nested regexes so I'm continuing to work on that.

markharwood committed:
Add options to rewrite to match none, match all or match all with verification.

markharwood committed:
Remove system.out call

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/ApproximateRegExp.java:
```diff
@@ -0,0 +1,631 @@
+                result = new TermQuery(new Term("", normalizedString));
+                break;
+            case REGEXP_CHAR:
+                String cs =Character.toString(c);
```
nit: 
```suggestion
                String cs = Character.toString(c);
```

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/ApproximateRegExp.java:
```diff
@@ -0,0 +1,631 @@
+            case REGEXP_INTERVAL:
+            case REGEXP_EMPTY: 
+            case REGEXP_AUTOMATON:
+                result = new MatchAllButRequireVerificationQuery();
```
Can we use a marker rather than a full Lucene query ?

markharwood replied:
See my earlier comment re introducing a base class or marker interface for denoting query clauses that require no verification step. Thoughts?

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/ApproximateRegExp.java:
```diff
@@ -0,0 +1,631 @@
+ * This class is a fork of Lucene's RegExp class and is used to create a simplified
+ * Query for use in accelerating searches to find those documents likely to match the regex. 
+ *
+ */
```
I don't know if it's a bad thing or not but creating boolean queries upfront removes the simplification of the automaton.

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,218 +231,383 @@ public WildcardFieldType clone() {
+                int clauseCount = 0;
+                for (BooleanClause clause : bq) {
+                    Query q = rewriteBoolToNgramQuery(clause.getQuery());
+                    if (q != null) {
```
When are we expecting null queries ? Should we also ignore `MatchAllQuery` since it's a valid query to return from `rewriteBoolToNgramQuery` ?

markharwood replied:
I think we can lose the null check.
MatchAllDocsQuery is intended to indicate that we have a `.*` type regex that needs no verification (unlike a `..` type regex where a MatchAllButRequireVerificationQuery indicates we need to just run the verification query).

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,218 +231,383 @@ public WildcardFieldType clone() {
+                // TODO we can revisit this approach at a later point if we want to introduce more like-for-like scoring with
+                // keyword field which will require tweaks to the verification query logic.
+                Automaton biggestEditDistanceAutomaton = automata[automata.length - 1].automaton;
+                bqBuilder.add(new AutomatonQueryOnBinaryDv(name(), searchTerm, biggestEditDistanceAutomaton), Occur.MUST);
```
The automaton can be big so we should try to store/create it in weights rather than queries (see https://issues.apache.org/jira/browse/LUCENE-9350).

jimczi commented on x-pack/plugin/wildcard/src/test/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapperTests.java:
```diff
@@ -221,8 +255,115 @@ public void testSearchResultsVersusKeywordField() throws IOException {
+        String patterns[] = { "*foobar", "foobar*", "foo*bar", "foo?bar", "?foo*bar?", "*c"};
+        for (String pattern : patterns) {
+            Query wildcardFieldQuery = wildcardFieldType.fieldType().wildcardQuery(pattern, null, MOCK_QSC);
+            assertTrue(wildcardFieldQuery instanceof BooleanQuery);
```
Thanks for adding the tests, can you do the same to test the logic of fuzzy query ?

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,218 +231,383 @@ public WildcardFieldType clone() {
+
+                BooleanQuery.Builder bqBuilder = new BooleanQuery.Builder();
 
+                // Add any prefixLength as a MUST clause to the main BooleanQuery
```
we could consider the entire prefix length instead of limiting to the first 3 characters ?

jimczi reviewed (commented):
I reviewed the logic for the regex, wildcard and fuzzy queries.
I am not convinced that the approximate regexp should use Lucene queries as intermediate states but you don't like visiting Automaton either so I left some comments on the approach.

markharwood commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/ApproximateRegExp.java:
```diff
@@ -0,0 +1,631 @@
+                    // TODO match all was a nice assumption here for optimising .* but breaks 
+                    // for (a){0,3} which isn't a logical match all but empty string or up to 3 a's. 
+//                    result = new MatchAllDocsQuery();
+                    result = new MatchAllButRequireVerificationQuery();
```
I thought this class made the intention much clearer. Without it we'd have to interpret any FieldExistsQuery to mean "don't run a verification" query  (e.g. for regex `.*`) and for a MatchAllDocsQuery to mean DO run a verification query (e.g. for regex `..`).
There are certain queries like `.*` where we know we can satisfy all criteria using the ngram index only.  However I thought today that we could extend this support by making MatchAllButRequireVerificationQuery a base class or marker interface. The regex `a` for example could also be satisfied using the ngram index only so returning a form of term query which tested for true in `instanceof MatchAllButRequireVerificationQuery` would help the WildcardField avoid running pointless verification for more than one regex type.

jimczi replied:
I see but is it still the case if the ngram index is lowercased ? The search is case-sensitive so unless we differentiate characters eligible for case folding, we have to check every match ?

markharwood replied:
We can see if the search value changes after normalisation and only optimise the query if there's no change.

jimczi replied:
> We can see if the search value changes after normalisation and only optimise the query if there's no change.
Perhaps a rare example of a user regex but illustrates the need to communicate where we think we can handle queries with the ngram index only.

I don't think this works since values in documents would be lowercased too so you'd need a verification for any character that can upper and lower cased ? We can think about optimizations in the future but as you said in previous comments we should aim for correctness first.

markharwood committed:
Addressing latest review comments by adding
* Unlimited prefix length.
* Delayed Autotmaton creation
* FuzzyQuery tests

markharwood committed:
Checkstyle fix

markharwood committed:
Checkstyle

markharwood commented:
It's worth noting the logic that this PR (and any other approach) has to implement.
It is more important to be correct than fast with this ngram acceleration logic - we can't run the risk of introducing false negatives. 
When it comes to converting regex clauses into ngram queries they fall into 3 categories:
A) Representable clauses we can search for e.g. `foo`
B) Verified match-all clauses e.g. `.*`
C) Unrepresentable clauses e.g. `...`

These clauses can be surrounded with arbitrary nesting of layers of AND/OR Boolean logic.
When it comes to avoiding false negatives the category C clauses are the ones we have to pay special attention to. They require us to rewrite any Boolean logic that surround these expressions using these rules (in order of precedence):
1) Any clause in an OR list which is a verified "match all" (e.g. `.*`) replaces all other clauses
2) Any clause in an OR list which is unrepresentable will rewrite the entire list to be a single unrepresentable clause
3) Any clause in an AND list which is unrepresentable can be dropped. An empty list is rewritten as an unrepresentable clause

I found implementing these rules easier to reason about using the hierarchy of RegExp/BooleanQuery objects rather than introducing an intermediate stage of an Automaton. The Automaton's graph of character-level transitions between states makes it harder to reason about the above Boolean rewriting logic.

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,218 +232,407 @@ public WildcardFieldType clone() {
                 TermQuery tq = new TermQuery(new Term(name(), token));
-                bqBuilder.add(new BooleanClause(tq, Occur.MUST));
+                bqBuilder.add(new BooleanClause(tq, occur));
             } else {
```
This shouldn't be possible ? Maybe add an assert that tokenSize should be less than or equals to NGRAM_SIZE at this point ?

markharwood replied:
Not sure I understand - do you mean the else is not possible? It should catch when tokenSize ==2. I think we agreed that `ab*` was worth running but not `a*`

>Maybe add an assert that tokenSize should be less than or equals to NGRAM_SIZE at this point ?

line 515 has this assert?

jimczi replied:
Sorry I missed the too short condition.

jimczi reviewed (commented):
I left more comments

markharwood committed:
Simplified ApproximateRegExp, removing dependency on WildcardFieldMapper class. All query simplification logic is now consolidated in the WildcardFieldMapper class.

markharwood committed:
Add test logic for @ syntax

markharwood committed:
Removed fork of Lucene’s RegEx now we have access to its internals

jimczi commented on x-pack/plugin/wildcard/build.gradle:
```diff
@@ -16,3 +16,9 @@ dependencies {
+licenseHeaders {
+  additionalLicense 'REGEX', 'RegExp (BSD-like)', 'Copyright (c) 2001-2009 Anders Moeller'
+  excludes << 'org/elasticsearch/xpack/wildcard/mapper/ApproximateRegExp.java'
+}
```
This is not needed anymore ?

jimczi commented on x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java:
```diff
@@ -216,218 +233,538 @@ public WildcardFieldType clone() {
+            String searchTerm = BytesRefs.toString(value);
+            String lowerSearchTerm = toLowerCase(searchTerm);
+            
+            try {
```
nit: Can you remove the empty lines ?

jimczi reviewed (approved):
I left minor comments but this looks ready to me. Thanks for iterating on this @markharwood!

markharwood committed:
Enabled integration tests, removed redundant license declaration and whitespace

markharwood committed:
Reverse last gradle change - can’t safely enable integration tests until this PR is backported to 7x
